adapters/mobilefuse/mobilefuse.go

```diff
@@ -0,0 +1,193 @@
+
+	for _, seatbid := range incomingBidResponse.SeatBid {
+		for i := range seatbid.Bid {
+			outgoingBidResponse.Bids = append(outgoingBidResponse.Bids, &adapters.TypedBid{
+				Bid:     &seatbid.Bid[i],
+				BidType: adapter.getBidType(seatbid.Bid[i].ImpID, incomingRequest.Imp),
+			})
+		}
+	}
+
+	return outgoingBidResponse, nil
+}
+
+func (adapter *MobilefuseAdapter) makeRequest(bidRequest *openrtb.BidRequest) (*adapters.RequestData, []error) {
+	var errs []error
+
+	mobilefuseExtension, errs := adapter.getMobilefuseExtension(bidRequest)
+
+	if errs != nil {
+		return nil, errs
+	}
+
+	endpoint, err := adapter.getEndpoint(mobilefuseExtension)
+
+	if err != nil {
+		return nil, append(errs, err)
+	}
+
+	validImps := adapter.getValidImps(bidRequest, mobilefuseExtension)
+
+	if len(validImps) == 0 {
+		err := fmt.Errorf("No valid imps")
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	mobilefuseBidRequest := *bidRequest
+	mobilefuseBidRequest.Imp = validImps
+	body, err := json.Marshal(mobilefuseBidRequest)
+
+	if err != nil {
+		return nil, append(errs, err)
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     endpoint,
+		Body:    body,
+		Headers: headers,
+	}, errs
+}
+
+func (adapter *MobilefuseAdapter) getMobilefuseExtension(request *openrtb.BidRequest) (*openrtb_ext.ExtImpMobilefuse, []error) {
+	var mobilefuseImpExtension openrtb_ext.ExtImpMobilefuse
+	var errs []error
+
+	for _, imp := range request.Imp {
+		var bidder_imp_extension adapters.ExtImpBidder
+
+		err := json.Unmarshal(imp.Ext, &bidder_imp_extension)
+
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		err = json.Unmarshal(bidder_imp_extension.Bidder, &mobilefuseImpExtension)
+
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		break
+	}
+
+	return &mobilefuseImpExtension, errs
+}
+
+func (adapter *MobilefuseAdapter) getEndpoint(ext *openrtb_ext.ExtImpMobilefuse) (string, error) {
+	publisher_id := strconv.Itoa(ext.PublisherId)
+
+	url, errs := macros.ResolveMacros(adapter.EndpointTemplate, macros.EndpointTemplateParams{PublisherID: publisher_id})
+
+	if errs != nil {
+		return "", errs
+	}
+
+	if ext.TagidSrc == "ext" {
+		url += "&tagid_src=ext"
+	}
+
+	return url, nil
+}
+
+func (adapter *MobilefuseAdapter) getValidImps(bidRequest *openrtb.BidRequest, ext *openrtb_ext.ExtImpMobilefuse) []openrtb.Imp {
+	var validImps []openrtb.Imp
+
+	for _, imp := range bidRequest.Imp {
+		if imp.Banner != nil || imp.Video != nil {
+			if imp.Banner != nil && imp.Video != nil {
+				imp.Video = nil
+			}
+
+			imp.TagID = strconv.Itoa(ext.PlacementId)
+			imp.Ext = nil
+			validImps = append(validImps, imp)
+
+			break
+		}
+	}
+
+	return validImps
+}
+
+func (adapter *MobilefuseAdapter) getBidType(imp_id string, imps []openrtb.Imp) openrtb_ext.BidType {
+	if imps[0].Video != nil {
```

Comment:
Is there a previous check which guarantees that the imps slice will never be empty?

Reply:
The request wouldn't be sent if there is no valid imp. See current line 98. There is also a unit test for this use case.

Reply:
You had me at "There is also a test for this use case." :)